[zh-tw] Translate content/zh-tw/docs/setup/production-environment/container_runtimes.md (#55253)
kubernetes/website: content/zh-tw/docs/setup/production-environment/container-runtimes.md

最重要的是，kubelet 和容器執行階段必須使用相同的 cgroup 驅動程式，且設定必須一致。

<!-- 
There are two cgroup drivers available:

* [`cgroupfs`](#cgroupfs-cgroup-driver)
* [`systemd`](#systemd-cgroup-driver)
-->
目前有兩種可用的 cgroup 驅動程式：

* [`cgroupfs`](#cgroupfs-cgroup-driver)
* [`systemd`](#systemd-cgroup-driver)

<!-- 
### cgroupfs driver {#cgroupfs-cgroup-driver}
-->
### cgroupfs 驅動程式 {#cgroupfs-cgroup-driver}

<!-- 
The `cgroupfs` driver is the [default cgroup driver in the kubelet](/docs/reference/config-api/kubelet-config.v1beta1).
 When the `cgroupfs` driver is used, the kubelet and the container runtime directly interface with
 the cgroup filesystem to configure cgroups.
-->
`cgroupfs` 驅動程式是 [kubelet 預設的 cgroup 驅動程式](/zh-tw/docs/reference/config-api/kubelet-config.v1beta1)。
當使用 `cgroupfs` 驅動程式時，kubelet 和容器執行階段會直接與 cgroup 檔案系統互動以設定 cgroup。

<!-- 
The `cgroupfs` driver is **not** recommended when
[systemd](https://www.freedesktop.org/wiki/Software/systemd/) is the
init system because systemd expects a single cgroup manager on
the system. Additionally, if you use [cgroup v2](/docs/concepts/architecture/cgroups), use the `systemd`
cgroup driver instead of `cgroupfs`.
-->

> 當 [systemd](https://www.freedesktop.org/wiki/Software/systemd/) 是系統的 init 系統時，**不建議**使用 `cgroupfs` 驅動程式，因為 systemd 預期系統上只有單一個 cgroup 管理器。此外，如果您使用 [cgroup v2](/zh-tw/docs/concepts/architecture/cgroups)，請使用 `systemd` cgroup 驅動程式而非 `cgroupfs`。

Reviewer: ```suggestion
當 [systemd](https://www.freedesktop.org/wiki/Software/systemd/) 是系統的 init 系統時，
**不建議**使用 `cgroupfs` 驅動程式，因為 systemd 預期系統上只有單一個 cgroup 管理器。
此外，如果您使用 [cgroup v2](/zh-tw/docs/concepts/architecture/cgroups)，請使用 `systemd` cgroup 驅動程式而非 `cgroupfs`。
```

---
title: 容器執行階段
content_type: concept
weight: 20
---
<!-- 
title: Container Runtimes
content_type: concept
weight: 20
-->

<!-- overview -->

{{% dockershim-removal %}}

<!-- 
You need to install a
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}
into each node in the cluster so that Pods can run there. This page outlines
what is involved and describes related tasks for setting up nodes.
-->
您需要在叢集中的每個節點安裝
{{< glossary_tooltip text="容器執行階段" term_id="container-runtime" >}}，

> 以便 Pod 能在其中執行。本頁概述了相關的內容，並說明設定節點的相關工作。

Reviewer 2: ```suggestion
以便 Pod 能在這些節點上執行。本頁概述所需的設定內容，並說明設定節點時的相關工作。
```

Kubernetes.

- [containerd](#containerd)
- [CRI-O](#cri-o)
- [Docker Engine](#docker)
- [Mirantis Container Runtime](#mcr)
-->
本頁概述了如何將數種常見的容器執行階段與 Kubernetes 搭配使用。

- [containerd](#containerd)
- [CRI-O](#cri-o)
- [Docker Engine](#docker)
- [Mirantis Container Runtime](#mcr)

{{< note >}}
<!--
Kubernetes releases before v1.24 included a direct integration with Docker Engine,
using a component named _dockershim_. That special direct integration is no longer
part of Kubernetes (this removal was
[announced](/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
as part of the v1.20 release).
You can read
[Check whether Dockershim removal affects you](/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-removal-affects-you/)
to understand how this removal might affect you. To learn about migrating from using dockershim, see
[Migrating from dockershim](/docs/tasks/administer-cluster/migrating-from-dockershim/).

If you are running a version of Kubernetes other than v{{< skew currentVersion >}},
check the documentation for that version.
-->
Kubernetes v1.24 之前的版本透過一個名為 _dockershim_ 的組件，直接與 Docker Engine 整合。

> 這種特殊的直接整合已不再是 Kubernetes 的一部分（此移除已在 v1.20 版本發布時[公告](/zh-tw/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)）。

Reviewer 2: ```suggestion
這種特殊的直接整合已不再是 Kubernetes 的一部分（這項移除已於 v1.20 版本發布時[公告](/zh-tw/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)）。
```

您可以閱讀[確認 Dockershim 移除是否會影響您](/zh-tw/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-removal-affects-you/)，
了解此移除可能對您造成的影響。
若要了解如何從 dockershim 遷移，請參閱[從 dockershim 遷移](/zh-tw/docs/tasks/administer-cluster/migrating-from-dockershim/)。

如果您執行的 Kubernetes 版本不是 v{{< skew currentVersion >}}，請查看該版本的文件。
{{< /note >}}

<!-- body -->
<!-- 
## Install and configure prerequisites
-->
## 安裝與設定前置作業

<!-- 
### Network configuration
-->
### 網路設定

<!-- 
By default, the Linux kernel does not allow IPv4 packets to be routed
between interfaces. Most Kubernetes cluster networking implementations
will change this setting (if needed), but some might expect the
administrator to do it for them. (Some might also expect other sysctl
parameters to be set, kernel modules to be loaded, etc; consult the
documentation for your specific network implementation.)
-->
預設情況下，Linux 核心不允許在介面之間路由 IPv4 封包。

> 大多數 Kubernetes 叢集網路實作都會修改此設定（如果需要），但有些實作可能會預期管理員代為執行此步驟。

Reviewer 2: ```suggestion
大多數 Kubernetes 叢集網路實作都會修改此設定（如果需要），但有些實作可能會要求管理員自行完成這項設定。
```

（有些實作也可能預期管理員設定其他 sysctl 參數、載入核心模組等；請針對您所使用的特定網路實作，參閱其說明文件。）

<!-- 
### Enable IPv4 packet forwarding {#prerequisite-ipv4-forwarding-optional}
-->
### 啟用 IPv4 封包轉發 {#prerequisite-ipv4-forwarding-optional}

<!-- 
To manually enable IPv4 packet forwarding:
-->
若要手動啟用 IPv4 封包轉發：

```bash

> # sysctl params required by setup, params persist across reboots

Reviewer 2: I think in general commands and configuration content inside code blocks should remain unchanged from upstream, but comments can be translated to help better understanding:

```suggestion
# 設定所需的 sysctl 參數，這些參數會在重新開機後保留
```

cat <<EOF | sudo tee /etc/sysctl.d/k8s.conf
net.ipv4.ip_forward = 1
EOF

> # Apply sysctl params without reboot

Reviewer 2: ```suggestion
# 不重新開機，直接套用 sysctl 參數
```

sudo sysctl --system
```

<!-- 
Verify that `net.ipv4.ip_forward` is set to 1 with:
-->
透過以下指令驗證 `net.ipv4.ip_forward` 是否已設定為 1：

```bash
sysctl net.ipv4.ip_forward
```

<!-- 
## cgroup drivers
-->
## cgroup 驅動程式

<!-- 
On Linux, {{< glossary_tooltip text="control groups" term_id="cgroup" >}}
are used to constrain resources that are allocated to processes.
-->

> 在 Linux 系統上，{{< glossary_tooltip text="控制群組" term_id="cgroup" >}} 是用於限制分配給程序的資源。

Reviewer 2: For consistency, suggest using "行程" for "process"
```suggestion
在 Linux 系統上，{{< glossary_tooltip text="控制群組" term_id="cgroup" >}} 是用於限制分配給行程的資源。
```

<!-- 
Both the {{< glossary_tooltip text="kubelet" term_id="kubelet" >}} and the
underlying container runtime need to interface with control groups to enforce
[resource management for pods and containers](/docs/concepts/configuration/manage-resources-containers/)
and set resources such as cpu/memory requests and limits. To interface with control
groups, the kubelet and the container runtime need to use a *cgroup driver*.
It's critical that the kubelet and the container runtime use the same cgroup
driver and are configured the same.
-->
{{< glossary_tooltip text="kubelet" term_id="kubelet" >}} 與底層的容器執行階段都需要與控制群組互動，
以強制執行 [Pod 與容器的資源管理](/zh-tw/docs/concepts/configuration/manage-resources-containers/)，並設定如 CPU、記憶體等資源的請求與限制。
為了與控制群組互動，kubelet 和容器執行階段需要使用 *cgroup 驅動程式*。

> 最重要的是，kubelet 和容器執行階段必須使用相同的 cgroup 驅動程式，且設定必須一致。

Reviewer 2: nit: maybe can try to make this sentence more written/formal, now is a bit colloquial

```suggestion
kubelet 和容器執行階段必須使用相同的 cgroup 驅動程式，且設定保持一致，這一點非常重要。
```

<!-- 
There are two cgroup drivers available:

* [`cgroupfs`](#cgroupfs-cgroup-driver)
* [`systemd`](#systemd-cgroup-driver)
-->

> 目前有兩種可用的 cgroup 驅動程式：

Reviewer 2: ```suggestion
有兩種可用的 cgroup 驅動程式：
```

* [`cgroupfs`](#cgroupfs-cgroup-driver)
* [`systemd`](#systemd-cgroup-driver)

<!-- 
### cgroupfs driver {#cgroupfs-cgroup-driver}
-->
### cgroupfs 驅動程式 {#cgroupfs-cgroup-driver}

<!-- 
The `cgroupfs` driver is the [default cgroup driver in the kubelet](/docs/reference/config-api/kubelet-config.v1beta1).
 When the `cgroupfs` driver is used, the kubelet and the container runtime directly interface with
 the cgroup filesystem to configure cgroups.
-->
`cgroupfs` 驅動程式是 [kubelet 預設的 cgroup 驅動程式](/zh-tw/docs/reference/config-api/kubelet-config.v1beta1)。

> 當使用 `cgroupfs` 驅動程式時，kubelet 和容器執行階段會直接與 cgroup 檔案系統互動以設定 cgroup。

Reviewer 2: ```suggestion
當使用 `cgroupfs` 驅動程式時，kubelet 和容器執行階段會直接透過 cgroup 檔案系統來設定 cgroup。
```

<!-- 
The `cgroupfs` driver is **not** recommended when
[systemd](https://www.freedesktop.org/wiki/Software/systemd/) is the
init system because systemd expects a single cgroup manager on
the system. Additionally, if you use [cgroup v2](/docs/concepts/architecture/cgroups), use the `systemd`
cgroup driver instead of `cgroupfs`.
-->
當 [systemd](https://www.freedesktop.org/wiki/Software/systemd/) 是系統的 init 系統時，

> **不建議**使用 `cgroupfs` 驅動程式，因為 systemd 預期系統上只有單一個 cgroup 管理器。

Reviewer 2: ```suggestion
**不建議**使用 `cgroupfs` 驅動程式，因為 systemd 預期系統上只有單一 cgroup 管理器。
```

此外，如果您使用 [cgroup v2](/zh-tw/docs/concepts/architecture/cgroups)，請使用 `systemd` cgroup 驅動程式而非 `cgroupfs`。

<!-- 
### systemd cgroup driver {#systemd-cgroup-driver}
-->
### systemd cgroup 驅動程式 {#systemd-cgroup-driver}

<!-- 
When [systemd](https://www.freedesktop.org/wiki/Software/systemd/) is chosen as the init
system for a Linux distribution, the init process generates and consumes a root control group
(`cgroup`) and acts as a cgroup manager.
-->
當 Linux 發行版選擇 [systemd](https://www.freedesktop.org/wiki/Software/systemd/) 作為 init 系統時，

> init 程序會產生並使用一個根控制群組（`cgroup`），並扮演 cgroup 管理器的角色。

Reviewer 2: Since "root control group" only appears once, and is referred to as "cgorup" afterwards, I suggest we keep the english term so that the abbreviated term is more obvious (while can keep chinese translation in brackets)

```suggestion
init 行程會產生並使用一個 root control group（根控制群組，`cgroup`），並扮演 cgroup 管理器的角色。
```

<!-- 
systemd has a tight integration with cgroups and allocates a cgroup per systemd
unit. As a result, if you use `systemd` as the init system with the `cgroupfs`
driver, the system gets two different cgroup managers.
-->
systemd 與 cgroup 具有緊密的整合，並會為每個 systemd 單元分配一個 cgroup。
因此，如果您將 `systemd` 作為 init 系統並搭配 `cgroupfs` 驅動程式使用，系統中將會出現兩個不同的 cgroup 管理器。

<!-- 
Two cgroup managers result in two views of the available and in-use resources in
the system. In some cases, nodes that are configured to use `cgroupfs` for the
kubelet and container runtime, but use `systemd` for the rest of the processes become
unstable under resource pressure.
-->
兩個 cgroup 管理器會導致系統對可用資源與已用資源產生兩種不同的解讀。

> 在某些情況下，如果節點被設定為在 kubelet 和容器執行階段使用 `cgroupfs`，而其他程序則使用 `systemd`；

Reviewer 2: ```suggestion
在某些情況下，如果節點被設定為在 kubelet 和容器執行階段使用 `cgroupfs`，而其他行程則使用 `systemd`；
```

> 當資源面臨壓力時，系統將會變得不穩定。

Reviewer 2: ```suggestion
在資源壓力下，系統可能會變得不穩定。
```

<!-- 
The approach to mitigate this instability is to use `systemd` as the cgroup driver for
the kubelet and the container runtime when systemd is the selected init system.
-->

> 緩解這種不穩定性的方法是，當 systemd 被選為 init 系統時，kubelet 和容器執行階段也應該使用 `systemd` 作為 cgroup 驅動程式。

Reviewer 2: ```suggestion
若 systemd 是所選的 init 系統，建議讓 kubelet 和容器執行階段也使用 `systemd` 作為 cgroup 驅動程式，以緩解這種不穩定性。
```

*  [containerd](#containerd-systemd)
*  [CRI-O](#cri-o)
-->
如果您將 kubelet 的 cgroup 驅動程式設定為 `systemd`，您也必須將容器執行階段的 cgroup 驅動程式設定為 `systemd`。相關說明請參閱您的容器執行階段文件。例如：

*  [containerd](#containerd-systemd)
*  [CRI-O](#cri-o)

<!-- 
In Kubernetes {{< skew currentVersion >}}, with the `KubeletCgroupDriverFromCRI`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
enabled and a container runtime that supports the `RuntimeConfig` CRI RPC,
the kubelet automatically detects the appropriate cgroup driver from the runtime,
and ignores the `cgroupDriver` setting within the kubelet configuration.
-->
在 Kubernetes {{< skew currentVersion >}} 中，若啟用 `KubeletCgroupDriverFromCRI`
[功能開關](/zh-tw/docs/reference/command-line-tools-reference/feature-gates/)，
並且容器執行階段支援 `RuntimeConfig` CRI RPC，kubelet 將自動從執行階段偵測合適的 cgroup 驅動程式，
並忽略 kubelet 設定中的 `cgroupDriver` 設定。

<!-- 
However, older versions of container runtimes (specifically,
containerd 1.y and below) do not support the `RuntimeConfig` CRI RPC, and
may not respond correctly to this query, and thus the Kubelet falls back to using the
value in its own `--cgroup-driver` flag.
-->

> 然而，較舊版本的容器執行階段（具體來說，containerd 1.y 及更低版本）不支援 `RuntimeConfig` CRI RPC，因此可能無法正確回應此查詢，導致 kubelet 退而使用自身 `--cgroup-driver` 參數的值。

Reviewer 2: ```suggestion
然而，較舊版本的容器執行階段（具體來說，containerd 1.y 及更低版本）不支援 `RuntimeConfig` CRI RPC，因此可能無法正確回應此查詢，導致 kubelet 改用自身 `--cgroup-driver` 參數的值。
```

<!-- 
In Kubernetes 1.36, this fallback behavior will be dropped, and older versions
of containerd will fail with newer kubelets.
-->

> 在 Kubernetes 1.36 中，此退回機制將被移除，這意味著較舊版本的 containerd 在較新的 kubelet 上將無法運作。

Reviewer 2: ```suggestion
在 Kubernetes 1.36 中，此退回機制將被移除，這意味著較舊版本的 containerd 將無法搭配較新的 kubelet 運作。
```

{{< caution >}}
<!--
Changing the cgroup driver of a Node that has joined a cluster is a sensitive operation.
If the kubelet has created Pods using the semantics of one cgroup driver, changing the container
runtime to another cgroup driver can cause errors when trying to re-create the Pod sandbox
for such existing Pods. Restarting the kubelet may not solve such errors.

If you have automation that makes it feasible, replace the node with another using the updated
configuration, or reinstall it using automation.
-->
對於已加入叢集的節點，變更其 cgroup 驅動程式是一項敏感的操作。

> 若 kubelet 已使用某種 cgroup 驅動程式的語意建立 Pod 的話，當嘗試重建這些既有 Pod 的沙盒時，將容器執行階段更換為另一種驅動程式可能會引發錯誤。這類錯誤可能無法透過重啟 kubelet 來解決。 

Reviewer 2: ```suggestion
若 kubelet 已依照某種 cgroup 驅動程式的語意建立 Pod，之後將容器執行階段改為另一種 cgroup 驅動程式時，嘗試重建這些既有 Pod 的沙盒可能會引發錯誤。這類錯誤可能無法透過重啟 kubelet 來解決。 
```

> 如果您的自動化機制可行，請使用已更新設定的另一個節點來替換該節點，或是透過自動化機制重新安裝該節點。

Reviewer 2: ```suggestion
如果您的自動化流程允許，請使用已更新設定的另一個節點來替換該節點，或是透過自動化機制重新安裝該節點。
```

{{< /caution >}}

<!-- 
### Migrating to the `systemd` driver in kubeadm managed clusters
-->
### 在 kubeadm 管理的叢集中遷移至 `systemd` 驅動程式

<!-- 
If you wish to migrate to the `systemd` cgroup driver in existing kubeadm managed clusters,
follow [configuring a cgroup driver](/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/).
-->
如果您希望在現有由 kubeadm 管理的叢集中遷移至 `systemd` cgroup 驅動程式，請遵循[設定 cgroup 驅動程式](/zh-tw/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/)的說明。

<!-- 
## CRI version support {#cri-versions}
-->
## CRI 版本支援 {#cri-versions}

<!-- 
Your container runtime must support at least v1alpha2 of the container runtime interface.
-->
您的容器執行階段必須至少支援 v1alpha2 的容器執行階段介面。

<!-- 
Kubernetes [starting v1.26](/blog/2022/11/18/upcoming-changes-in-kubernetes-1-26/#cri-api-removal)
_only works_ with v1 of the CRI API. Earlier versions default
to v1 version, however if a container runtime does not support the v1 API, the kubelet falls back to
using the (deprecated) v1alpha2 API instead.
-->
Kubernetes [自 v1.26 起](/zh-tw/blog/2022/11/18/upcoming-changes-in-kubernetes-1-26/#cri-api-removal)

> _僅支援_ v1 版本的 CRI API。較早版本預設使用 v1 版本，但如果容器執行階段不支援 v1 API，kubelet 將退而使用（已棄用）v1alpha2 API。

Reviewer 2: ```suggestion
_僅支援_ v1 版本的 CRI API。較早版本預設使用 v1 版本，但如果容器執行階段不支援 v1 API，kubelet 將改用（已棄用的）v1alpha2 API。
```

    SystemdCgroup = true
```

<!-- 
The `systemd` cgroup driver is recommended if you use [cgroup v2](/docs/concepts/architecture/cgroups).
-->
如果您使用 [cgroup v2](/zh-tw/docs/concepts/architecture/cgroups)，則建議使用 `systemd` cgroup 驅動程式。

{{< note >}}
<!--
If you installed containerd from a package (for example, RPM or `.deb`), you may find
that the CRI integration plugin is disabled by default.

You need CRI support enabled to use containerd with Kubernetes. Make sure that `cri`
is not included in the`disabled_plugins` list within `/etc/containerd/config.toml`;
if you made changes to that file, also restart `containerd`.

If you experience container crash loops after the initial cluster installation or after
installing a CNI, the containerd configuration provided with the package might contain
incompatible configuration parameters. Consider resetting the containerd configuration
with `containerd config default > /etc/containerd/config.toml` as specified in
[getting-started.md](https://github.com/containerd/containerd/blob/main/docs/getting-started.md#advanced-topics)
and then set the configuration parameters specified above accordingly.
-->

> 如果您是透過套件（例如 RPM 或 `.deb`）安裝 containerd，您可能會發現 CRI 整合外掛程式預設是被停用的。

Reviewer 2: ```suggestion
如果您是透過套件（例如 RPM 或 `.deb`）安裝 containerd，您可能會發現 CRI 整合外掛程式預設為停用。
```

> 您需要啟用 CRI 支援才能讓 containerd 與 Kubernetes 一起運作。請確保 `cri` 未被包含在 `/etc/containerd/config.toml` 中的 `disabled_plugins` 列表中；如果您修改了該檔案，請同時重新啟動 `containerd`。

Reviewer 2: ```suggestion
您需要啟用 CRI 支援才能讓 containerd 與 Kubernetes 一起運作。請確認 `/etc/containerd/config.toml` 中的 `disabled_plugins` 列表未包含 `cri`；如果您修改了該檔案，請同時重新啟動 `containerd`。
```

> 如果您在叢集初始安裝後或安裝 CNI 後遇到容器持續當機，可能是因為套件提供的 containerd 設定包含不相容的參數。建議參照 [getting-started.md](https://github.com/containerd/containerd/blob/main/docs/getting-started.md#advanced-topics) 的說明，使用 `containerd config default > /etc/containerd/config.toml` 重設 containerd 設定，再根據上述說明設定相關參數。

Reviewer 2: ```suggestion
如果您在叢集初始安裝後或安裝 CNI 後遇到容器不斷重啟，可能是因為套件提供的 containerd 設定包含不相容的參數。建議參照 [getting-started.md](https://github.com/containerd/containerd/blob/main/docs/getting-started.md#advanced-topics) 的說明，使用 `containerd config default > /etc/containerd/config.toml` 重設 containerd 設定，再依照上方說明設定指定的參數。
```

{{< /note >}}

<!-- 
If you apply this change, make sure to restart containerd:
-->

> 如果您套用了此變更，請務必重新啟動 containerd：

Reviewer 2: ```suggestion
如果您套用了上述變更，請務必重新啟動 containerd：
```

```shell
sudo systemctl restart containerd
```

<!-- 
When using kubeadm, manually configure the
[cgroup driver for kubelet](/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/#configuring-the-kubelet-cgroup-driver).
-->
使用 kubeadm 時，請手動設定 [kubelet 的 cgroup 驅動程式](/zh-tw/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/#configuring-the-kubelet-cgroup-driver)。

<!-- 
In Kubernetes v1.28, you can enable automatic detection of the
cgroup driver as an alpha feature. See [systemd cgroup driver](#systemd-cgroup-driver)
for more details.
-->

> 在 Kubernetes v1.28 中，您可以將自動偵測 cgroup 驅動程式啟用為 alpha 功能。如需更多細節，請參閱 [systemd cgroup 驅動程式](#systemd-cgroup-driver)。

Reviewer 2: ```suggestion
在 Kubernetes v1.28 中，您可以將自動偵測 cgroup 驅動程式啟用為 alpha 階段功能。如需更多細節，請參閱 [systemd cgroup 驅動程式](#systemd-cgroup-driver)。
```

<!-- 
#### Overriding the sandbox (pause) image {#override-pause-image-containerd}
-->

> #### 覆寫沙盒（pause）映像檔 {#override-pause-image-containerd}

Reviewer 2: ```suggestion
#### 覆寫沙盒的 pause 映像檔 {#override-pause-image-containerd}
```

pause refers to the pause image used for the Pod sandbox, not an English translation of "沙盒". Keeping pause next to "映像檔" avoids the possible misreading of "沙盒（pause）" as “sandbox = pause”.

<!-- 
### CRI-O
-->
### CRI-O

<!-- 
This section contains the necessary steps to install CRI-O as a container runtime.
-->
本節包含將 CRI-O 安裝為容器執行階段所需的步驟。

<!-- 
To install CRI-O, follow [CRI-O Install Instructions](https://github.com/cri-o/packaging/blob/main/README.md#usage).
-->
若要安裝 CRI-O，請遵循 [CRI-O 安裝說明](https://github.com/cri-o/packaging/blob/main/README.md#usage)。

<!-- 
#### cgroup driver
-->
#### cgroup 驅動程式

<!-- 
CRI-O uses the systemd cgroup driver per default, which is likely to work fine
for you. To switch to the `cgroupfs` cgroup driver, either edit
`/etc/crio/crio.conf` or place a drop-in configuration in
`/etc/crio/crio.conf.d/02-cgroup-manager.conf`, for example:
-->
CRI-O 預設使用 systemd cgroup 驅動程式，這通常可以正常運作。若要切換為 `cgroupfs` cgroup 驅動程式，您可以編輯 `/etc/crio/crio.conf` 或是在 `/etc/crio/crio.conf.d/02-cgroup-manager.conf` 放置一個補充設定檔，例如：

```toml
[crio.runtime]
conmon_cgroup = "pod"
cgroup_manager = "cgroupfs"
```

<!-- 
You should also note the changed `conmon_cgroup`, which has to be set to the value
`pod` when using CRI-O with `cgroupfs`. It is generally necessary to keep the
cgroup driver configuration of the kubelet (usually done via kubeadm) and CRI-O
in sync.
-->

> 您也應注意已變更的 `conmon_cgroup`。當 CRI-O 與 `cgroupfs` 搭配使用時，`conmon_cgroup` 必須被設定為 `pod`。一般來說，必須保持 kubelet（通常透過 kubeadm 設定）和 CRI-O 的 cgroup 驅動程式設定同步。

Reviewer 2: ```suggestion
您也應注意已變更的 `conmon_cgroup`。當 CRI-O 與 `cgroupfs` 搭配使用時，`conmon_cgroup` 必須設定為 `pod`。一般來說，必須保持 kubelet（通常透過 kubeadm 設定）和 CRI-O 的 cgroup 驅動程式設定同步。
```

<!-- 
In Kubernetes v1.28, you can enable automatic detection of the
cgroup driver as an alpha feature. See [systemd cgroup driver](#systemd-cgroup-driver)
for more details.
-->

> 在 Kubernetes v1.28 中，您可以將自動偵測 cgroup 驅動程式啟用為 alpha 功能。如需更多細節，請參閱 [systemd cgroup 驅動程式](#systemd-cgroup-driver)。

Reviewer 2: ```suggestion
在 Kubernetes v1.28 中，您可以將自動偵測 cgroup 驅動程式啟用為 alpha 階段功能。如需更多細節，請參閱 [systemd cgroup 驅動程式](#systemd-cgroup-driver)。
```

<!-- 
For CRI-O, the CRI socket is `/var/run/crio/crio.sock` by default.
-->
對於 CRI-O，預設的 CRI socket 是 `/var/run/crio/crio.sock`。

<!-- 
#### Overriding the sandbox (pause) image {#override-pause-image-cri-o}
-->

> #### 覆寫沙盒（pause）映像檔 {#override-pause-image-cri-o}

Reviewer 2: ```suggestion
#### 覆寫沙盒的 pause 映像檔 {#override-pause-image-cri-o}
```

<!-- 
In your [CRI-O config](https://github.com/cri-o/cri-o/blob/main/docs/crio.conf.5.md) you can set the following
config value:
-->
在您的 [CRI-O 設定](https://github.com/cri-o/cri-o/blob/main/docs/crio.conf.5.md)中，您可以設定以下配置值：

```toml
[crio.image]
pause_image="registry.k8s.io/pause:3.10"
```

<!-- 
This config option supports live configuration reload to apply this change: `systemctl reload crio` or by sending
`SIGHUP` to the `crio` process.
-->

> 此設定選項支援即時重新載入以套用變更：執行 `systemctl reload crio`，或向 `crio` 程序發送 `SIGHUP` 訊號。

Reviewer 2: ```suggestion
此設定選項支援即時重新載入以套用變更：執行 `systemctl reload crio`，或向 `crio` 行程發送 `SIGHUP` 訊號。
```

<!-- 
### Docker Engine {#docker}
-->
### Docker Engine {#docker}

{{< note >}}
<!--
These instructions assume that you are using the
[`cri-dockerd`](https://mirantis.github.io/cri-dockerd/) adapter to integrate
Docker Engine with Kubernetes.
-->
這些說明假設您正在使用 [`cri-dockerd`](https://mirantis.github.io/cri-dockerd/) 轉接器，以將 Docker Engine 與 Kubernetes 整合。
{{< /note >}}

<!-- 
1. On each of your nodes, install Docker for your Linux distribution as per
  [Install Docker Engine](https://docs.docker.com/engine/install/#server).
-->
1. 在您的每個節點上，根據[安裝 Docker Engine](https://docs.docker.com/engine/install/#server) 的說明，為您的 Linux 發行版安裝 Docker。

<!-- 
2. Install [`cri-dockerd`](https://mirantis.github.io/cri-dockerd/usage/install), following the directions in the install section of the documentation.
-->

> 2. 安裝 [`cri-dockerd`](https://mirantis.github.io/cri-dockerd/usage/install)，請遵循文件安裝章節中的說明。

Reviewer 2: ```suggestion
2. 請依照文件中安裝章節的說明，安裝 [`cri-dockerd`](https://mirantis.github.io/cri-dockerd/usage/install)。
```

<!-- 
For `cri-dockerd`, the CRI socket is `/run/cri-dockerd.sock` by default.
-->
對於 `cri-dockerd`，預設的 CRI socket 是 `/run/cri-dockerd.sock`。

<!-- 
### Mirantis Container Runtime {#mcr}
-->
### Mirantis 容器執行階段 {#mcr}

<!-- 
[Mirantis Container Runtime](https://docs.mirantis.com/mcr/25.0/overview.html) (MCR) is a commercially
available container runtime that was formerly known as Docker Enterprise Edition.
-->
[Mirantis 容器執行階段](https://docs.mirantis.com/mcr/25.0/overview.html) (MCR) 是一種商用的容器執行階段，其前身為 Docker Enterprise Edition。

<!-- 
You can use Mirantis Container Runtime with Kubernetes using the open source
[`cri-dockerd`](https://mirantis.github.io/cri-dockerd/) component, included with MCR.
-->
您可以透過 MCR 內建的開源 [`cri-dockerd`](https://mirantis.github.io/cri-dockerd/) 組件，將 Mirantis 容器執行階段搭配 Kubernetes 使用。

<!-- 
To learn more about how to install Mirantis Container Runtime,
visit [MCR Deployment Guide](https://docs.mirantis.com/mcr/25.0/install.html).
-->

> 要進一步了解如何安裝 Mirantis 容器執行階段，請訪問 [MCR 部署指南](https://docs.mirantis.com/mcr/25.0/install.html)。

Reviewer 2: ```suggestion
要進一步了解如何安裝 Mirantis 容器執行階段，請參閱 [MCR 部署指南](https://docs.mirantis.com/mcr/25.0/install.html)。
```

<!-- 
Check the systemd unit named `cri-docker.socket` to find out the path to the CRI
socket.
-->
檢查名為 `cri-docker.socket` 的 systemd 單元，以找出 CRI socket 的路徑。

<!-- 
#### Overriding the sandbox (pause) image {#override-pause-image-cri-dockerd-mcr}
-->

> #### 覆寫沙盒（pause）映像檔 {#override-pause-image-cri-dockerd-mcr}

Reviewer 2: ```suggestion
#### 覆寫沙盒的 pause 映像檔 {#override-pause-image-cri-dockerd-mcr}
```

<!-- 
The `cri-dockerd` adapter accepts a command line argument for
specifying which container image to use as the Pod infrastructure container (“pause image”).
The command line argument to use is `--pod-infra-container-image`.
-->

> `cri-dockerd` 轉接器接受一個命令列參數，用於指定要使用哪個容器映像檔作為 Pod 的基礎設施容器（"pause 映像檔"）。

Reviewer 2: ```suggestion
`cri-dockerd` 轉接器接受一個命令列參數，用於指定要作為 Pod 基礎設施容器（「pause 映像檔」）使用的容器映像檔。
```

要使用的命令列參數為 `--pod-infra-container-image`。

<!-- 
## {{% heading "whatsnext" %}}
-->
## {{% heading "whatsnext" %}}

<!-- 
As well as a container runtime, your cluster will need a working
[network plugin](/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model).
-->

> 除了容器執行階段外，您的叢集還需要一個可用的[網路外掛程式](/zh-tw/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model)。

Reviewer 2: ```suggestion
除了容器執行階段之外，您的叢集也需要可用的[網路外掛程式](/zh-tw/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model)。
```
